# Procesamiento de Resultados de Entrenamiento

Este notebook lee todos los archivos `.pkl` de resultados de diferentes seeds y los combina en un único archivo `global_results` similar al formato original.

In [1]:
import pickle
import os
from pathlib import Path
import re

## Configuración

Especifica la profundidad que quieres procesar:

In [2]:
depth = "in_3_4"  # Opciones: "in_0_1", "in_1_2", "in_2_3", "in_3_4", etc.

# Directorio base de resultados
base_dir = Path(f"training_results/depth_{depth}")

print(f"Directorio base: {base_dir}")
print(f"Buscando archivos .pkl en subdirectorios...")

Directorio base: training_results/depth_in_3_4
Buscando archivos .pkl en subdirectorios...


## Función para leer todos los archivos .pkl

In [3]:
def load_all_pkl_files(base_directory):
    """
    Lee todos los archivos .pkl en las subcarpetas de la profundidad especificada.
    
    Args:
        base_directory: Path object del directorio base (ej: training_results/depth_in_0_1)
    
    Returns:
        global_results: diccionario con formato {seed: seed_results}
    """
    global_results = {}
    
    # Verificar que el directorio existe
    if not base_directory.exists():
        print(f"Error: El directorio {base_directory} no existe")
        return global_results
    
    # Buscar todas las subcarpetas que empiezan con "seed"
    seed_dirs = sorted([d for d in base_directory.iterdir() if d.is_dir() and d.name.startswith('seed')])
    
    if not seed_dirs:
        print(f"No se encontraron carpetas de seeds en {base_directory}")
        return global_results
    
    print(f"Encontradas {len(seed_dirs)} carpetas de seeds")
    print("="*70)
    
    # Procesar cada carpeta de seed
    for seed_dir in seed_dirs:
        # Extraer el número de seed del nombre de la carpeta
        seed_match = re.search(r'seed(\d+)', seed_dir.name)
        if not seed_match:
            print(f"Saltando {seed_dir.name}: no se pudo extraer el número de seed")
            continue
        
        seed = int(seed_match.group(1))
        
        # Buscar archivos .pkl en esta carpeta
        pkl_files = list(seed_dir.glob('*.pkl'))
        
        if not pkl_files:
            print(f"Seed {seed}: No se encontraron archivos .pkl")
            continue
        
        # Si hay múltiples archivos, tomar el primero (normalmente solo hay uno)
        pkl_file = pkl_files[0]
        
        try:
            # Cargar el archivo pickle
            with open(pkl_file, 'rb') as f:
                seed_results = pickle.load(f)
            
            # Agregar a global_results
            global_results[seed] = seed_results
            
            # Información sobre lo que se cargó
            num_configs = len(seed_results) if isinstance(seed_results, dict) else 0
            print(f"Seed {seed:5d}: {num_configs} configuraciones cargadas desde {pkl_file.name}")
            
        except Exception as e:
            print(f"Seed {seed}: Error al cargar {pkl_file.name}: {e}")
    
    print("="*70)
    print(f"\n Total de seeds procesadas: {len(global_results)}")
    
    return global_results

## Cargar todos los resultados

In [4]:
global_results = load_all_pkl_files(base_dir)

Encontradas 100 carpetas de seeds
Seed 10639: 100 configuraciones cargadas desde full_results_in_3_4.pkl
Seed 10680: 100 configuraciones cargadas desde full_results_in_3_4.pkl
Seed 10860: 100 configuraciones cargadas desde full_results_in_3_4.pkl
Seed 10893: 100 configuraciones cargadas desde full_results_in_3_4.pkl
Seed 10922: 100 configuraciones cargadas desde full_results_in_3_4.pkl
Seed 10942: 100 configuraciones cargadas desde full_results_in_3_4.pkl
Seed 11282: 100 configuraciones cargadas desde full_results_in_3_4.pkl
Seed 11653: 100 configuraciones cargadas desde full_results_in_3_4.pkl
Seed 11826: 100 configuraciones cargadas desde full_results_in_3_4.pkl
Seed 12015: 100 configuraciones cargadas desde full_results_in_3_4.pkl
Seed 12273: 100 configuraciones cargadas desde full_results_in_3_4.pkl
Seed 12768: 100 configuraciones cargadas desde full_results_in_3_4.pkl
Seed  1281: 100 configuraciones cargadas desde full_results_in_3_4.pkl
Seed 13545: 100 configuraciones cargadas de

## Inspeccionar los resultados cargados

In [5]:
# Ver las seeds disponibles
print(f"Seeds cargadas: {sorted(global_results.keys())}")
print(f"\n Total de seeds: {len(global_results)}")

# Ver un ejemplo de estructura de datos (primera seed)
if global_results:
    first_seed = sorted(global_results.keys())[0]
    print(f"\n Estructura de datos de seed {first_seed}:")
    print(f"   Tipo: {type(global_results[first_seed])}")
    
    if isinstance(global_results[first_seed], dict):
        sample_keys = list(global_results[first_seed].keys())[:5]
        print(f"   Configuraciones totales: {len(global_results[first_seed])}")
        print(f"   Primeras 5 configuraciones (dataset, modelo):")
        for key in sample_keys:
            print(f"      {key}")

Seeds cargadas: [638, 679, 859, 892, 921, 941, 1281, 1652, 1825, 2014, 2272, 2767, 3545, 3707, 4024, 4330, 4388, 4434, 4456, 4503, 4504, 5003, 5132, 5264, 5454, 5545, 6316, 6438, 6545, 6779, 6973, 7005, 7174, 7357, 7580, 7611, 7739, 7783, 7815, 7860, 8227, 8276, 8397, 8582, 8585, 8878, 8931, 9267, 9706, 9756, 10639, 10680, 10860, 10893, 10922, 10942, 11282, 11653, 11826, 12015, 12273, 12768, 13545, 13708, 14024, 14330, 14389, 14434, 14457, 14504, 14505, 15004, 15132, 15265, 15454, 15546, 16317, 16439, 16546, 16779, 16973, 17005, 17175, 17357, 17581, 17611, 17739, 17784, 17815, 17860, 18227, 18276, 18397, 18583, 18586, 18879, 18931, 19267, 19707, 19756]

 Total de seeds: 100

 Estructura de datos de seed 638:
   Tipo: <class 'dict'>
   Configuraciones totales: 100
   Primeras 5 configuraciones (dataset, modelo):
      ('TOA_9x9_depth_in_3_4', 'XGB')
      ('TOA_9x9_depth_in_3_4', 'LBM')
      ('TOA_9x9_depth_in_3_4', 'MLP')
      ('TOA_9x9_depth_in_3_4', 'SVR')
      ('TOA_9x9_depth_in_

## Guardar resultados globales consolidados

In [6]:
# Crear el directorio si no existe
base_dir.mkdir(parents=True, exist_ok=True)

# Nombre del archivo de salida
output_file = base_dir / f"global_results_{depth}.pkl"

try:
    with open(output_file, 'wb') as f:
        pickle.dump(global_results, f)
    
    print(f"\n{'='*70}")
    print(f"RESULTADOS GLOBALES GUARDADOS EXITOSAMENTE")
    print(f"{'='*70}")
    print(f"Archivo: {output_file}")
    print(f"Seeds incluidas: {len(global_results)}")
    print(f"Tamaño del archivo: {output_file.stat().st_size / (1024*1024):.2f} MB")
    
except Exception as e:
    print(f"\nError al guardar resultados globales: {e}")


RESULTADOS GLOBALES GUARDADOS EXITOSAMENTE
Archivo: training_results/depth_in_3_4/global_results_in_3_4.pkl
Seeds incluidas: 100
Tamaño del archivo: 72.25 MB


## Verificar el archivo guardado

In [7]:
# Recargar el archivo para verificar que se guardó correctamente
try:
    with open(output_file, 'rb') as f:
        verified_results = pickle.load(f)
    
    print("Verificación exitosa:")
    print(f"Seeds en archivo guardado: {len(verified_results)}")
    print(f"Seeds originales: {len(global_results)}")
    print(f"¿Coinciden?: {'Sí ✓' if len(verified_results) == len(global_results) else 'No ✗'}")
    
except Exception as e:
    print(f"❌ Error al verificar el archivo: {e}")

Verificación exitosa:
Seeds en archivo guardado: 100
Seeds originales: 100
¿Coinciden?: Sí ✓


## Análisis rápido de los resultados (opcional)

Si quieres ver estadísticas básicas de los resultados:

In [8]:
# Contar configuraciones únicas
all_configs = set()
for seed, seed_results in global_results.items():
    if isinstance(seed_results, dict):
        all_configs.update(seed_results.keys())

print("Estadísticas generales:")
print(f"Total de seeds: {len(global_results)}")
print(f"Configuraciones únicas (dataset, modelo): {len(all_configs)}")

# Ver distribución de datasets
datasets = set([config[0] for config in all_configs if isinstance(config, tuple)])
print(f"\n Datasets encontrados: {len(datasets)}")
for dataset in sorted(datasets):
    print(f"   - {dataset}")

# Ver distribución de modelos
models = set([config[1] for config in all_configs if isinstance(config, tuple)])
print(f"\n Modelos encontrados: {len(models)}")
for model in sorted(models):
    print(f"   - {model}")

Estadísticas generales:
Total de seeds: 100
Configuraciones únicas (dataset, modelo): 100

 Datasets encontrados: 10
   - C2X-Complex_rhow_15x15_depth_in_3_4
   - C2X-Complex_rhow_5x5_depth_in_3_4
   - C2X-Complex_rhow_9x9_depth_in_3_4
   - C2X-Complex_rhown_5x5_depth_in_3_4
   - C2X-Complex_rhown_9x9_depth_in_3_4
   - TOA_15x15_depth_in_3_4
   - TOA_1x1_depth_in_3_4
   - TOA_3x3_depth_in_3_4
   - TOA_5x5_depth_in_3_4
   - TOA_9x9_depth_in_3_4

 Modelos encontrados: 10
   - CAT
   - ELN
   - ENSEMBLE
   - KNN
   - LBM
   - LR
   - MLP
   - RF
   - SVR
   - XGB


In [9]:
import pandas as pd

# Seleccionar una seed (puedes cambiar esto)
if global_results:
    selected_seed = sorted(global_results.keys())[0]  # Primera seed disponible
    print(f"Mostrando resultados de seed {selected_seed}")
    
    # Extraer datos para el DataFrame
    rows = []
    
    for (dataset, model), results in global_results[selected_seed].items():
        row = {
            'Dataset': dataset,
            'Modelo': model,
        }
        
        # Añadir métricas de entrenamiento
        if 'train_metrics' in results:
            for metric, value in results['train_metrics'].items():
                row[f'Train_{metric}'] = value
        
        # Añadir métricas de test
        if 'test_metrics' in results:
            for metric, value in results['test_metrics'].items():
                row[f'Test_{metric}'] = value
        
        rows.append(row)
    
    # Crear DataFrame
    df = pd.DataFrame(rows)
    
    print(f"\n DataFrame con {len(df)} configuraciones")
    print(f"Columnas: {list(df.columns)}")
    
    # Mostrar el DataFrame
    display(df)
    
    # Opcional: guardar como CSV
    # df.to_csv(f'results_seed{selected_seed}.csv', index=False)
    # print(f"\n DataFrame guardado como results_seed{selected_seed}.csv")
else:
    print(" No hay resultados cargados")

Mostrando resultados de seed 638

 DataFrame con 100 configuraciones
Columnas: ['Dataset', 'Modelo', 'Train_RMSLE', 'Train_RMSE', 'Train_R2', 'Test_RMSLE', 'Test_RMSE', 'Test_R2']


,Dataset,Modelo,Train_RMSLE,Train_RMSE,Train_R2,Test_RMSLE,Test_RMSE,Test_R2
0,TOA_9x9_depth_in_3_4,XGB,0.189297,1.740280,0.689926,0.373598,1.933573,0.519418
1,TOA_9x9_depth_in_3_4,LBM,0.200052,0.585418,0.964912,0.410396,1.948889,0.511774
2,TOA_9x9_depth_in_3_4,MLP,0.314979,1.130560,0.869138,0.438876,2.029496,0.470552
3,TOA_9x9_depth_in_3_4,SVR,0.363395,1.857924,0.646586,0.435360,2.157327,0.401756
4,TOA_9x9_depth_in_3_4,KNN,0.004968,0.005763,0.999997,0.440085,2.105152,0.430343
...,...,...,...,...,...,...,...,...
95,C2X-Complex_rhown_9x9_depth_in_3_4,LR,0.508071,2.248840,0.467195,0.450817,2.144317,0.380487
96,C2X-Complex_rhown_9x9_depth_in_3_4,RF,0.332480,1.312279,0.818573,0.411585,2.069908,0.422736
97,C2X-Complex_rhown_9x9_depth_in_3_4,CAT,0.229188,0.677719,0.951611,0.387716,1.802426,0.562289
98,C2X-Complex_rhown_9x9_depth_in_3_4,ELN,0.503217,2.236380,0.473083,0.445999,2.067522,0.424066


In [12]:
"""
Código ACTUALIZADO para generar DataFrames de métricas con media ± desviación estándar
- Métricas: R2, RMSE, RMSLE
- Ordenamiento de datasets según el mejor modelo (como en Entrenamiento_Final.ipynb)
"""

import pandas as pd
import numpy as np

# Asegurarse de que global_results esté cargado
if not global_results:
    print("Error: global_results no está cargado. Ejecuta primero las celdas anteriores.")
else:
    print(f"Procesando resultados de {len(global_results)} seeds\n")
    
    # Definir las métricas que queremos extraer
    metrics = ['R2', 'RMSE', 'RMSLE']
    splits = ['train', 'test']
    
    # Diccionario para almacenar los dataframes
    dfs_dict = {}
    
    # Procesar cada métrica y split
    for metric in metrics:
        for split in splits:
            print(f"Procesando {metric} ({split})...")
            
            # Diccionario temporal para almacenar valores por dataset y modelo
            # Estructura: {(dataset, modelo): [valores de todas las seeds]}
            values_dict = {}
            
            # Iterar sobre todas las seeds
            for seed, seed_results in global_results.items():
                for (dataset, model), results in seed_results.items():
                    # Obtener la métrica correspondiente
                    metrics_key = f'{split}_metrics'
                    if metrics_key in results and metric in results[metrics_key]:
                        value = results[metrics_key][metric]
                        
                        # Inicializar lista si no existe
                        if (dataset, model) not in values_dict:
                            values_dict[(dataset, model)] = []
                        
                        values_dict[(dataset, model)].append(value)
            
            print(f"   Encontradas {len(values_dict)} configuraciones únicas")
            
            # Verificar si hay datos
            if len(values_dict) == 0:
                print(f"   No se encontraron datos para {metric} ({split})")
                continue
            
            # Crear DataFrame con media ± std
            rows = []
            for (dataset, model), values in values_dict.items():
                mean = np.mean(values)
                std = np.std(values)
                rows.append({
                    'Dataset': dataset,
                    'Modelo': model,
                    f'{metric}_{split}': f'{mean:.3f} ± {std:.3f}'
                })
            
            # Verificar que tenemos datos antes de pivotear
            if len(rows) == 0:
                print(f"   No se pudieron crear filas para {metric} ({split})")
                continue
            
            # Crear DataFrame y pivotear
            df_temp = pd.DataFrame(rows)
            df_pivot = df_temp.pivot(index='Dataset', columns='Modelo', values=f'{metric}_{split}')
            
            # Renombrar columnas para que coincidan con el formato esperado
            # El formato esperado tiene ENS en lugar de ENSEMBLE
            if 'ENSEMBLE' in df_pivot.columns:
                df_pivot.rename(columns={'ENSEMBLE': 'ENS'}, inplace=True)
            
            # Ordenar columnas alfabéticamente
            df_pivot = df_pivot.reindex(sorted(df_pivot.columns), axis=1)
            
            # Guardar en el diccionario (se ordenarán las filas después)
            key = f'{metric}_{split}'
            dfs_dict[key] = df_pivot
            
            print(f"   DataFrame creado: {df_pivot.shape[0]} datasets × {df_pivot.shape[1]} modelos\n")
    
    # ORDENAR DATASETS según el criterio de la función create_df_results
    print("Ordenando datasets según mejor modelo...")
    
    for metric in metrics:
        # Para R2: ordenar por máximo valor de test (descendente)
        if metric == 'R2' and f'{metric}_test' in dfs_dict:
            df_test = dfs_dict[f'{metric}_test'].copy()
            df_train = dfs_dict[f'{metric}_train'].copy()
            
            # Calcular el máximo R2 test para cada dataset (ignorando formato "X.XX ± X.XX")
            max_values = df_test.apply(lambda row: max([float(val.split(' ± ')[0]) for val in row]), axis=1)
            sorted_indices = max_values.sort_values(ascending=False).index
            
            # Reordenar ambos DataFrames
            dfs_dict[f'{metric}_test'] = df_test.loc[sorted_indices]
            dfs_dict[f'{metric}_train'] = df_train.loc[sorted_indices]
            print(f"   {metric}: ordenado por max test (descendente)")
        
        # Para RMSE y RMSLE: ordenar por mínimo valor de test (ascendente)
        elif metric in ['RMSE', 'RMSLE'] and f'{metric}_test' in dfs_dict:
            df_test = dfs_dict[f'{metric}_test'].copy()
            df_train = dfs_dict[f'{metric}_train'].copy()
            
            # Calcular el mínimo para cada dataset
            min_values = df_test.apply(lambda row: min([float(val.split(' ± ')[0]) for val in row]), axis=1)
            sorted_indices = min_values.sort_values(ascending=True).index
            
            # Reordenar ambos DataFrames
            dfs_dict[f'{metric}_test'] = df_test.loc[sorted_indices]
            dfs_dict[f'{metric}_train'] = df_train.loc[sorted_indices]
            print(f"   {metric}: ordenado por min test (ascendente)")
    
    print(f"\n{'='*70}")
    print(f"DATAFRAMES CREADOS Y ORDENADOS EXITOSAMENTE")
    print(f"{'='*70}")
    print(f"\nDataFrames disponibles:")
    for key in sorted(dfs_dict.keys()):
        print(f"   - dfs_dict['{key}']")
    
    # Guardar los DataFrames en CSV
    print(f"\nGuardando DataFrames en CSV...")
    
    # Crear directorio de salida
    output_dir = Path(f"training_results/depth_{depth}/results_csv")
    output_dir.mkdir(parents=True, exist_ok=True)
    
    for key, df in sorted(dfs_dict.items()):
        # Nombre del archivo
        output_file = output_dir / f"results_{depth}_{key}.csv"
        
        # Guardar CSV
        df.to_csv(output_file)
        
        print(f"   ✓ {output_file.name}")
    
    print(f"\nArchivos guardados en: {output_dir}")
    print(f"\n{'='*70}")
    print(f"PROCESO COMPLETADO")
    print(f"{'='*70}")

Procesando resultados de 100 seeds

Procesando R2 (train)...
   Encontradas 100 configuraciones únicas
   DataFrame creado: 10 datasets × 10 modelos

Procesando R2 (test)...
   Encontradas 100 configuraciones únicas
   DataFrame creado: 10 datasets × 10 modelos

Procesando RMSE (train)...
   Encontradas 100 configuraciones únicas
   DataFrame creado: 10 datasets × 10 modelos

Procesando RMSE (test)...
   Encontradas 100 configuraciones únicas
   DataFrame creado: 10 datasets × 10 modelos

Procesando RMSLE (train)...
   Encontradas 100 configuraciones únicas
   DataFrame creado: 10 datasets × 10 modelos

Procesando RMSLE (test)...
   Encontradas 100 configuraciones únicas
   DataFrame creado: 10 datasets × 10 modelos

Ordenando datasets según mejor modelo...
   R2: ordenado por max test (descendente)
   RMSE: ordenado por min test (ascendente)
   RMSLE: ordenado por min test (ascendente)

DATAFRAMES CREADOS Y ORDENADOS EXITOSAMENTE

DataFrames disponibles:
   - dfs_dict['R2_test']
   - 

In [11]:
"""
Código para encontrar el mejor modelo-dataset
Prioridad: RMSLE > R2 > RMSE
Score combinado: 50% RMSLE + 10% desv RMSLE + 35% R2 + 5% desv R2
"""

import pandas as pd
import numpy as np

print("ANÁLISIS DE MEJORES MODELOS-DATASETS\n")
print("Orden de prioridad: RMSLE > R2 > RMSE\n")
print("Score: 50% RMSLE + 10% desv.RMSLE + 35% R2 + 5% desv.R2\n")
print("="*70)

if 'dfs_dict' not in locals():
    print("Error: dfs_dict no está disponible. Ejecuta primero la celda anterior.")
else:
    # ========== 1. MEJOR RMSLE TEST (PRIORIDAD ALTA) ==========
    print("\n MEJOR RMSLE TEST (Métrica prioritaria):")
    df_rmsle_test = dfs_dict['RMSLE_test']
    
    # Encontrar el mínimo global
    min_values = df_rmsle_test.apply(lambda row: min([float(val.split(' ± ')[0]) for val in row]), axis=1)
    best_rmsle_dataset = min_values.idxmin()
    best_rmsle_model = df_rmsle_test.loc[best_rmsle_dataset].apply(lambda x: float(x.split(' ± ')[0])).idxmin()
    
    print(f"   Dataset: {best_rmsle_dataset}")
    print(f"   Modelo: {best_rmsle_model}")
    print(f"   RMSLE Test: {df_rmsle_test.loc[best_rmsle_dataset, best_rmsle_model]}")
    
    # Mostrar también las otras métricas para este par
    print(f"   R2 Test: {dfs_dict['R2_test'].loc[best_rmsle_dataset, best_rmsle_model]}")
    print(f"   RMSE Test: {dfs_dict['RMSE_test'].loc[best_rmsle_dataset, best_rmsle_model]}")
    
    # Verificar overfitting
    print(f"   RMSLE Train: {dfs_dict['RMSLE_train'].loc[best_rmsle_dataset, best_rmsle_model]}")
    print(f"   R2 Train: {dfs_dict['R2_train'].loc[best_rmsle_dataset, best_rmsle_model]}")
    
    # ========== 2. TOP 5 MEJORES RMSLE ==========
    print(f"\n{'='*70}")
    print(" TOP 5 MEJORES POR RMSLE TEST:\n")
    
    # Crear lista de todos los pares (dataset, modelo) con sus valores
    rmsle_results = []
    for dataset in df_rmsle_test.index:
        for model in df_rmsle_test.columns:
            # RMSLE
            rmsle_str = df_rmsle_test.loc[dataset, model]
            rmsle_mean = float(rmsle_str.split(' ± ')[0])
            rmsle_std = float(rmsle_str.split(' ± ')[1])
            
            # R2
            r2_str = dfs_dict['R2_test'].loc[dataset, model]
            r2_mean = float(r2_str.split(' ± ')[0])
            r2_std = float(r2_str.split(' ± ')[1])
            
            rmsle_results.append({
                'Dataset': dataset,
                'Modelo': model,
                'RMSLE_Test': rmsle_mean,
                'RMSLE_Std': rmsle_std,
                'R2_Test': r2_mean,
                'R2_Std': r2_std,
                'RMSLE_Test_Full': rmsle_str,
                'R2_Test_Full': r2_str
            })
    
    # Ordenar por RMSLE (ascendente)
    df_ranked = pd.DataFrame(rmsle_results).sort_values('RMSLE_Test')
    
    # Mostrar top 5
    for idx, row in df_ranked.head(5).iterrows():
        rank = list(df_ranked.head(5).index).index(idx) + 1
        print(f"{rank}. {row['Dataset']} + {row['Modelo']}")
        print(f"   RMSLE Test: {row['RMSLE_Test_Full']}")
        print(f"   R2 Test: {row['R2_Test_Full']}")
        print()
    
    # ========== 3. ANÁLISIS DE ESTABILIDAD ==========
    print(f"{'='*70}")
    print(" TOP 5 MÁS ESTABLES (menor desviación estándar en RMSLE):\n")
    
    df_stable = pd.DataFrame(rmsle_results).sort_values('RMSLE_Std')
    
    for idx, row in df_stable.head(5).iterrows():
        rank = list(df_stable.head(5).index).index(idx) + 1
        print(f"{rank}. {row['Dataset']} + {row['Modelo']}")
        print(f"   RMSLE Test: {row['RMSLE_Test_Full']} (Desv. Std: {row['RMSLE_Std']:.3f})")
        print(f"   R2 Test: {row['R2_Test_Full']} (Desv. Std: {row['R2_Std']:.3f})")
        print()
    
    # ========== 4. CRITERIO COMBINADO: RMSLE + R2 + Estabilidad ==========
    print(f"{'='*70}")
    print(" RECOMENDACIÓN FINAL (Score combinado):\n")
    print("   Pesos: 50% RMSLE + 10% desv.RMSLE + 35% R2 + 5% desv.R2\n")
    
    # Score combinado con nuevos pesos
    df_analysis = pd.DataFrame(rmsle_results)
    
    # Normalizar RMSLE (0-1, menor es mejor)
    rmsle_normalized = (df_analysis['RMSLE_Test'] - df_analysis['RMSLE_Test'].min()) / \
                       (df_analysis['RMSLE_Test'].max() - df_analysis['RMSLE_Test'].min())
    
    # Normalizar desviación de RMSLE (0-1, menor es mejor)
    rmsle_std_normalized = (df_analysis['RMSLE_Std'] - df_analysis['RMSLE_Std'].min()) / \
                           (df_analysis['RMSLE_Std'].max() - df_analysis['RMSLE_Std'].min())
    
    # Normalizar R2 (0-1, mayor es mejor, así que invertimos)
    r2_normalized = 1 - ((df_analysis['R2_Test'] - df_analysis['R2_Test'].min()) / \
                         (df_analysis['R2_Test'].max() - df_analysis['R2_Test'].min()))
    
    # Normalizar desviación de R2 (0-1, menor es mejor)
    r2_std_normalized = (df_analysis['R2_Std'] - df_analysis['R2_Std'].min()) / \
                        (df_analysis['R2_Std'].max() - df_analysis['R2_Std'].min())
    
    # Score combinado (menor es mejor)
    # 50% RMSLE + 10% desv.RMSLE + 35% R2 (invertido) + 5% desv.R2
    df_analysis['Combined_Score'] = (0.50 * rmsle_normalized + 
                                      0.10 * rmsle_std_normalized + 
                                      0.35 * r2_normalized + 
                                      0.05 * r2_std_normalized)
    
    # Ordenar por mejor score (menor es mejor)
    df_final = df_analysis.sort_values('Combined_Score')
    
    for idx, row in df_final.head(3).iterrows():
        rank = list(df_final.head(3).index).index(idx) + 1
        print(f"{rank}. {row['Dataset']} + {row['Modelo']}")
        print(f"   RMSLE Test: {row['RMSLE_Test_Full']}")
        print(f"   R2 Test: {row['R2_Test_Full']}")
        print(f"   Score Combinado: {row['Combined_Score']:.3f} (menor es mejor)")
        print()
    
    print(f"{'='*70}")
    print("\n  CONCLUSIÓN:")
    print(f"   El mejor par modelo-dataset es:")
    print(f"   {df_final.iloc[0]['Dataset']} + {df_final.iloc[0]['Modelo']}")
    print(f"   Con RMSLE Test = {df_final.iloc[0]['RMSLE_Test_Full']}")
    print(f"   Y R2 Test = {df_final.iloc[0]['R2_Test_Full']}")

ANÁLISIS DE MEJORES MODELOS-DATASETS

Orden de prioridad: RMSLE > R2 > RMSE

Score: 50% RMSLE + 10% desv.RMSLE + 35% R2 + 5% desv.R2


 MEJOR RMSLE TEST (Métrica prioritaria):
   Dataset: TOA_15x15_depth_in_3_4
   Modelo: CAT
   RMSLE Test: 0.392 ± 0.035
   R2 Test: 0.598 ± 0.092
   RMSE Test: 1.889 ± 0.342
   RMSLE Train: 0.058 ± 0.030
   R2 Train: 0.997 ± 0.003

 TOP 5 MEJORES POR RMSLE TEST:

1. TOA_15x15_depth_in_3_4 + CAT
   RMSLE Test: 0.392 ± 0.035
   R2 Test: 0.598 ± 0.092

2. TOA_9x9_depth_in_3_4 + CAT
   RMSLE Test: 0.396 ± 0.033
   R2 Test: 0.587 ± 0.091

3. TOA_15x15_depth_in_3_4 + KNN
   RMSLE Test: 0.397 ± 0.041
   R2 Test: 0.565 ± 0.118

4. TOA_15x15_depth_in_3_4 + ENS
   RMSLE Test: 0.406 ± 0.036
   R2 Test: 0.587 ± 0.083

5. TOA_9x9_depth_in_3_4 + KNN
   RMSLE Test: 0.406 ± 0.041
   R2 Test: 0.556 ± 0.115

 TOP 5 MÁS ESTABLES (menor desviación estándar en RMSLE):

1. TOA_1x1_depth_in_3_4 + ENS
   RMSLE Test: 0.440 ± 0.030 (Desv. Std: 0.030)
   R2 Test: 0.526 ± 0.082 (D

#### 0 - 1

   **C2X-Complex_rhow_15x15_depth_in_0_1 + CAT**

   - Con RMSLE Test = 0.337 ± 0.036

   - Y R2 Test = 0.755 ± 0.098

#### 1 - 2

   **C2X-Complex_rhow_15x15_depth_in_1_2 + CAT**

   - Con RMSLE Test = 0.365 ± 0.048
   
   - Y R2 Test = 0.759 ± 0.097

#### 2 - 3

   **TOA_15x15_depth_in_2_3 + CAT**
   
   - Con RMSLE Test = 0.378 ± 0.035
   
   - Y R2 Test = 0.723 ± 0.091

#### 3 - 4

   **TOA_15x15_depth_in_3_4 + CAT**
   
   - Con RMSLE Test = 0.392 ± 0.035
   
   - Y R2 Test = 0.598 ± 0.092